lib/parser/canonicalize-omp.cc

```diff
@@ -0,0 +1,126 @@
+  template<typename T> void Post(T &) {}
+  CanonicalizationOfOmp(Messages &messages) : messages_{messages} {}
+
+  void Post(Block &block) {
+    for (auto it{block.begin()}; it != block.end(); ++it) {
+      if (auto *exec{std::get_if<ExecutableConstruct>(&it->u)}) {
+        if (auto *ompCons{
+                std::get_if<common::Indirection<OpenMPConstruct>>(&exec->u)}) {
+          // OpenMPLoopConstruct
+          if (auto *ompLoop{
+                  std::get_if<OpenMPLoopConstruct>(&ompCons->value().u)}) {
+            RewriteOpenMPLoopConstruct(ompLoop, block, it);
+          }
+        } else if (auto *endDir{
+                       std::get_if<common::Indirection<OpenMPEndLoopDirective>>(
+                           &exec->u)}) {
+          // Unmatched OpenMPEndloopdirective
+          messages_.Say(endDir->value().source,
+              "The %s directive must follow the DO loop associated with the "
+              "loop construct"_err_en_US,
+              parser::ToUpperCaseLetters(endDir->value().source.ToString()));
+        }
+      }
+    }  // Block list
+  }
+
+private:
+  template<typename T> T *GetConstructIf(parser::ExecutionPartConstruct &x) {
+    if (auto *y{std::get_if<ExecutableConstruct>(&x.u)}) {
+      if (auto *z{std::get_if<common::Indirection<T>>(&y->u)}) {
+        return &z->value();
+      }
+    }
+    return nullptr;
+  }
+
+  void RewriteOpenMPLoopConstruct(
+      OpenMPLoopConstruct *x, Block &block, Block::iterator it) {
+    // Check the sequence of DoConstruct and OpenMPEndLoopDirective
+    // in the same iteration
+    //
+    // Original:
+    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct
+    //     OmpLoopDirective
+    //     OmpClauseList
+    //   ExecutableConstruct -> DoConstruct
+    //   ExecutableConstruct -> OpenMPEndLoopDirective (if available)
+    //
+    // After rewriting:
+    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct
+    //     OmpLoopDirective
+    //     OmpClauseList
+    //     DoConstruct
+    //     OpenMPEndLoopDirective (if available)
+    Block::iterator nextIt;
+    auto &dir{std::get<OmpLoopDirective>(x->t)};
+
+    nextIt = it;
+    if (++nextIt != block.end()) {
+      if (auto *doCons{GetConstructIf<DoConstruct>(*nextIt)}) {
+        const auto &loopControl{doCons->GetLoopControl()};
+        if (loopControl.has_value()) {
+          // move DoConstruct
+          std::get<std::optional<DoConstruct>>(x->t) = std::move(*doCons);
+          nextIt = block.erase(nextIt);
+
+          // try to match OpenMPEndLoopDirective
+          if (auto *endDir{GetConstructIf<OpenMPEndLoopDirective>(*nextIt)}) {
+            std::get<std::optional<OpenMPEndLoopDirective>>(x->t) =
+                std::move(*endDir);
+            nextIt = block.erase(nextIt);
+          }
+        } else {
+          messages_.Say(dir.source,
+              "DO loop after the %s directive "
+              "must have loop control"_err_en_US,
```

Comment:
This message can probably fit on one line now (because the indentation is reduced). That's always better because it makes it easier to grep for error messages.